### PR TITLE
Add cli API for CLIPlugin and also changes to reflect the changes don…

### PR DESCRIPTION
**Pull request by prkalle**
…e related to k8s API dependency in tanzu-plugin-runtime module

- Add cli API for CLIPlugin (migrate apis/cli module from tanzu-framework to tanzu-cli)
- Changes to reflect the changes done related to removal of k8s API dependency in tanzu-plugin-runtime module

Signed-off-by: Prem Kumar Kalle <pkalle@vmware.com>

This PR migrates the CLI APIs from tanzu-framework to tanzu-cli. It also has changes necessary to reflect the k8s API dependency changes done in tanzu-plugin-runtime module.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR
Tested CLI with legacy cluster plugin(using the framework runtime dependency) and also with test plugin(using the latest plugin-runtime) and it was success

```
❯ ./artifacts/darwin/arm64/cli/core/dev/tanzu-cli-darwin_arm64 plugin sync
ℹ  Checking for required plugins... 
ℹ  Installing plugin 'kubernetes-release:v0.28.0-dev' with target 'kubernetes'
ℹ  Installing plugin 'isolated-cluster:v0.28.0-dev'
ℹ  Installing plugin 'login:v0.28.0-dev'
ℹ  Installing plugin 'codegen:v0.28.0-dev'
ℹ  Installing plugin 'cluster:v0.28.0-dev' with target 'kubernetes'
ℹ  Installing plugin 'feature:v0.28.0-dev' with target 'kubernetes'
ℹ  Installing plugin 'pinniped-auth:v0.28.0-dev'
ℹ  Installing plugin 'secret:v0.28.0-dev' with target 'kubernetes'
ℹ  Installing plugin 'management-cluster:v0.28.0-dev' with target 'kubernetes'
ℹ  Installing plugin 'builder:v0.28.0-dev'
ℹ  Installing plugin 'test:v0.28.0-dev'
ℹ  Installing plugin 'helloworld:v0.28.0-dev' with target 'kubernetes'
ℹ  Installing plugin 'package:v0.28.0-dev' with target 'kubernetes'
ℹ  Installing plugin 'telemetry:v0.28.0-dev' with target 'kubernetes'
ℹ  Successfully installed all required plugins 
✔  Done 
❯ ./artifacts/darwin/arm64/cli/core/dev/tanzu-cli-darwin_arm64 helloworld print
Warning, Masking commands for plugins "login" because a core command or other plugin with that name already exists. 
Hello world from test pluginCurrent context - &{Name:tkg-aws-cc-capi115-upg-mc Target:kubernetes GlobalOpts:<nil> ClusterOpts:0x140001495c0 DiscoverySources:[]}%                                                                       
❯ 
❯ ./artifacts/darwin/arm64/cli/core/dev/tanzu-cli-darwin_arm64 cluster list
Warning, Masking commands for plugins "login" because a core command or other plugin with that name already exists. 
  NAME                            NAMESPACE  STATUS    CONTROLPLANE  WORKERS  KUBERNETES         ROLES   PLAN  TKR                                   
  tkg-aws-cc-capi115-upg-wc       default    updating  1/1           1/1      v1.23.8+vmware.1   <none>        v1.23.8---vmware.1-tkg.1-zshippable   
  tkg-aws-cc-capi115-upg-wc-test  default    running   1/1           1/1      v1.21.14+vmware.2  <none>        v1.21.14---vmware.2-tkg.1-zshippable  
```

### Release note

```release-note

```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -249,6 +250,21 @@ generate-fakes: ## Generate fakes for writing unit tests
...
+CONTROLLER_GEN_SRC ?= "./..."
+
+generate-controller-code: $(CONTROLLER_GEN)  ## Generate code via controller-gen
+	$(CONTROLLER_GEN) $(GENERATOR) object:headerFile="$(ROOT_DIR)/hack/boilerplate.go.txt",year=$(shell date +%Y) paths="$(CONTROLLER_GEN_SRC)" $(OPTIONS)
+	$(MAKE) fmt
+
+generate-manifests:
+	$(MAKE) generate-controller-code GENERATOR=crd OPTIONS="output:crd:artifacts:config=$(ROOT_DIR)/apis/config/crd/bases" CONTROLLER_GEN_SRC=$(CONTROLLER_GEN_SRC)
+
+generate: generate-controller-code generate-manifests
```
Comment (anujc25) [inline Makefile:266]: nit: Add help text to make it appear in the `make help` output.

```suggestion
generate: generate-controller-code generate-manifests ## Generate controller code and manifests e.g. CRD etc.
```

File: Makefile [hunk 2]
```diff
@@ -249,6 +250,21 @@ generate-fakes: ## Generate fakes for writing unit tests
...
+## Generators
+## --------------------------------------
+
+CONTROLLER_GEN_SRC ?= "./..."
+
+generate-controller-code: $(CONTROLLER_GEN)  ## Generate code via controller-gen
+	$(CONTROLLER_GEN) $(GENERATOR) object:headerFile="$(ROOT_DIR)/hack/boilerplate.go.txt",year=$(shell date +%Y) paths="$(CONTROLLER_GEN_SRC)" $(OPTIONS)
+	$(MAKE) fmt
+
+generate-manifests:
```
Comment (anujc25) [inline Makefile:263]: nit: Add help text to make it appear in the `make help` output.

```suggestion
generate-manifests: ## Generate API manifests e.g. CRD
```

**Review by anujc25 — COMMENTED** [review]
Overall changes look good.

However, When running `make generate-manifests` locally, it changed the formatting in `apis/config/crd/bases/cli.tanzu.vmware.com_cliplugins.yaml` file. Can you please double check to verify the behavior on your system? Also, we should catch this in the CI if someone forgets to update the CRD when updating the golang API code.

```
-    - name: v1alpha1
-      schema:
-        openAPIV3Schema:
-          description: CLIPlugin denotes a Tanzu cli plugin.
-          properties:
-            apiVersion:
-              description: 'APIVersion defines the versioned schema of this representation
+  - name: v1alpha1
+    schema:
+      openAPIV3Schema:
+        description: CLIPlugin denotes a Tanzu cli plugin.
+        properties:
+          apiVersion:
+            description: 'APIVersion defines the versioned schema of this representation
```

File: Makefile (hunk 1 above)
Comment (prkalle) [reply Makefile:266]: updated. thanks!

**Comment by prkalle** [comment]
> Overall changes look good.
> 
> However, When running `make generate-manifests` locally, it changed the formatting in `apis/config/crd/bases/cli.tanzu.vmware.com_cliplugins.yaml` file. Can you please double check to verify the behavior on your system?

Thanks for catching this. It's my IDE  that did yaml formatting which seems off from the yaml that being genereated by controller-gen. 

>Also, we should catch this in the CI if someone forgets to update the CRD when updating the golang API code.

Updated the make `verify` target to catch the missing CRD updates

File: Makefile [hunk 3]
```diff
@@ -247,8 +248,24 @@ generate-fakes: ## Generate fakes for writing unit tests
 
 .PHONY: verify
 verify: gomod fmt ## Run all verification scripts
+	$(MAKE)  generate
```
Comment (anujc25) [inline Makefile:251]:
```suggestion
verify: gomod fmt generate ## Run all verification scripts
```

**Review by anujc25 — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**